core/src/multi_proof.rs

```diff
@@ -0,0 +1,548 @@
+        // When there are more than two key_paths, the same logic is applied,
+        // but if two key_paths have different bits at an index, the key_paths are split
+        // based on the value at that index.
+        //
+        // If all key_paths share a bit at an index, that sibling is required;
+        // if at least two key_paths differ, no sibling is needed, but the key_paths must be divided
+        // based on having bit 0 or 1 at that index.
+        //
+        // Iterate this algorithm on the bisection to determine the minimum necessary siblings.
+        //
+        // `external_siblings` whill follow this structure for each bicection:
+        // |commons siblings| ext siblings in the right bisection | ext siblings in the left bisection |
+
+        // it represent a slice of the vector of key_paths with the notion of
+        // path_index, at which bit in the key_path that slice is pointing at
+        struct TraverseInfo {
+            path_index: usize,
+            path_slice_lower: usize,
+            path_slice_upper: usize,
+            ext_sibling_index: usize,
+        }
+        let mut stack_traverse_info: Vec<TraverseInfo> = vec![];
+
+        // External siblings encountered during a subtree traversal
+        let mut ext_sibling_in_sub_tree: Vec<Node> = vec![];
+
+        // initially we're looking at all the key_paths
+        let mut curr_traverse_info = TraverseInfo {
+            path_index: 0,
+            path_slice_lower: 0,
+            path_slice_upper: path_proofs.len() - 1,
+            ext_sibling_index: 0,
+        };
+
+        loop {
+            // Based on how we bisect the key_path slice, we can see that each TraverseInfo
+            // will represent a slice where all key_paths will have same bits before
+            // TraverseInfo.path_index.
+
+            // If there is no longer a common prefix, there is only one key in the TraverseInfo,
+            // and then all remaining siblings are required for the multiproof.
+            if curr_traverse_info.path_slice_lower == curr_traverse_info.path_slice_upper {
+                let path_proof = &path_proofs[curr_traverse_info.path_slice_lower];
+                let sub_path_proof = SubPathProof {
+                    terminal: path_proof.terminal.clone(),
+                    // The depth at which the terminal starts not sharing any siblings
+                    depth: curr_traverse_info.path_index as u8,
+                    inner_siblings: path_proof
+                        .siblings
+                        .iter()
+                        .skip(curr_traverse_info.path_index)
+                        .copied()
+                        .collect(),
+                };
+                sub_paths.push(sub_path_proof);
+
+                // insert all collected siblings in this sub tree in ext_sibling
+                // at the position this bisection is pointing at
+                for sibling in ext_sibling_in_sub_tree.into_iter().rev() {
```

Comment:
`ext_sibling_in_sub_tree` is guaranteed empty here because terminal nodes always follow immediately after internal nodes. you can omit this code.

Reply:
Ok, you're totally right in saying that `ext_sibling_in_sub_tree` is guaranteed to be empty here; however, I didn't understand your explanation.

What I'm thinking is that if we have found a unique sub-path, then we can be sure that there can't be `ext_sibling_in_sub_tree` because they were already emptied in the last bisection that brought us to this unique sub-path